src/EFCore.PG/Query/Sql/Internal/NpgsqlQuerySqlGenerator.cs

```diff
@@ -189,54 +188,52 @@ protected override Expression VisitBinary(BinaryExpression expression)
                 goto default;
             }
 
-            case ExpressionType.ArrayIndex:
-                VisitArrayIndex(expression);
-                return expression;
-
             default:
                 return base.VisitBinary(expression);
             }
         }
 
         /// <inheritdoc />
-        protected override Expression VisitUnary(UnaryExpression expression)
-        {
-            if (expression.NodeType == ExpressionType.ArrayLength)
-            {
-                VisitSqlFunction(new SqlFunctionExpression("array_length", typeof(int), new[] { expression.Operand, Expression.Constant(1) }));
-                return expression;
-            }
-
-            return base.VisitUnary(expression);
-        }
-
-        protected virtual void VisitArrayIndex([NotNull] BinaryExpression expression)
+        protected override Expression VisitIndex(IndexExpression expression)
         {
-            Debug.Assert(expression.NodeType == ExpressionType.ArrayIndex);
-
-            if (expression.Left.Type == typeof(byte[]))
-            {
-                // bytea cannot be subscripted, but there's get_byte
-                VisitSqlFunction(new SqlFunctionExpression("get_byte", typeof(byte),
-                    new[] { expression.Left, expression.Right }));
-                return;
-            }
-
-            if (expression.Left.Type == typeof(string))
+            // bytea cannot be subscripted
+            if (expression.Object.Type == typeof(byte[]))
+                return
+                    VisitSqlFunction(
+                        new SqlFunctionExpression(
+                            "get_byte",
+                            typeof(byte),
+                            new[] { expression.Object, expression.Arguments[0] }));
+
+            // text cannot be subscripted
+            if (expression.Object.Type == typeof(string))
+                return
+                    VisitSqlFunction(
+                        new SqlFunctionExpression(
+                            "ascii",
```

Comment:
C# string subscripting yields a char, which the EF Core provider maps to `text` - so why convert to an int here via `ascii()`?

Regardless, it would be better to not include this kind of change in such a big PR, it's very easy to miss while reviewing...

Reply:
We discussed this (a while back now) in #450. I haven't found a good solution for identifying when `byte` is really `char`, so I was going to stick with this... But I'm ___very___ open to a better solution, this one just feels so hackish.

Reply:
I was able to identify this case in `VisitUnary` and apply `ascii` there and out of the indexers.